allow setting default deployment mode for Kserve in DSC

VedantMahabaleshwarkar opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This change is to support Raw Deployments for Kserve. JIRA: https://issues.redhat.com/browse/RHOAIENG-3132 
This PR adds a field `DefaultDeploymentMode` to the Kserve spec in the DSC. The default value is `Serverless`, which introduces no changes in behavior. However, if the `DefaultDeploymentMode` is set to `RawDeployment`, the operator updates the configmap `inferenceservice-config` in the applications NS. The updates are as follows : 
- deploy["defaultDeploymentMode"] is set to match the value passed to the DSC since they both use the same enum
- ingress["disableIngressCreation "] is set to `true`  if the deployment mode is `RawDeployments`, otherwise it is `false` 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Changes tested against dev cluster by building and deploying `quay.io/vedantm/opendatahub-operator:rawdeployment` 
Testing instructions : 
- clone the operator repo and run `make deploy -e IMG=quay.io/vedantm/opendatahub-operator:latest -e OPERATOR_NAMESPACE=opendatahub-operator-system` 
- Create a DSCInitialization as follows
```
spec:
  applicationsNamespace: opendatahub
  monitoring:
    managementState: Managed
    namespace: opendatahub
  serviceMesh:
    controlPlane:
      metricsCollection: Istio
      name: data-science-smcp
      namespace: istio-system
    managementState: Unmanaged
``` 
- Create a DSC as follows 
```
    kserve:
      defaultDeploymentMode: RawDeployment
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: SelfSigned
        managementState: Removed
        name: knative-serving
``` 
- Once the DSC reconcile is successful, verify values in the `inferenceservice-config` ConfigMap in the applications NS 
![image](https://github.com/opendatahub-io/opendatahub-operator/assets/20891020/2ca18c13-cd36-49f5-814a-bfe61ca5781c)
![image](https://github.com/opendatahub-io/opendatahub-operator/assets/20891020/770cf72d-4db5-4174-9f6d-d25c490b7aad)
- Change the DSC spec such that default deployment is now `Serverless`, wait for DSC to reconcile and verify the corresponding changes in the `inferenceservice-config` object 



## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

terrytangyuan commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
+	inferenceServiceConfigMap := &corev1.ConfigMap{}
+	err := cli.Get(ctx, client.ObjectKey{
+		Namespace: dscispec.ApplicationsNamespace,
+		Name:      "inferenceservice-config",
```
Move to constant

terrytangyuan commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
+	var ingressData map[string]interface{}
+	json.Unmarshal([]byte(inferenceServiceConfigMap.Data["ingress"]), &ingressData)
+	if k.DefaultDeploymentMode == RawDeployment {
+		ingressData["ingressClassName"] = RawDeploymentIngressClassName
```
Move to constant

VedantMahabaleshwarkar replied:
removed the logic for `IngressClassName` based on your PR to allow disableIngressCreation

terrytangyuan commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
+
+	//if the default deployment mode is RawDeployment, also set the ingressClass to "openshift-default", otherwise it should be "istio"
+	var ingressData map[string]interface{}
+	json.Unmarshal([]byte(inferenceServiceConfigMap.Data["ingress"]), &ingressData)
```
Missing error handling

terrytangyuan commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
+	var deployData map[string]interface{}
+	json.Unmarshal([]byte(inferenceServiceConfigMap.Data["deploy"]), &deployData)
+	deployData["defaultDeploymentMode"] = k.DefaultDeploymentMode
+	deployDataBytes, _ := json.MarshalIndent(deployData, "", " ")
```
Missing error handling

bartoszmajsak replied:
Thanks for spotting it @terrytangyuan. Interestingly enough we have linter gh action which reports that too https://github.com/opendatahub-io/opendatahub-operator/actions/runs/7894716777/job/21545878156?pr=864#step:5:33, though I was expecting to comment back on the PR instead of just dumping it in the logs.

I am lacking enough karma but I think what we are missing is `write` permission for this GH Action https://github.com/golangci/golangci-lint-action#comments-and-annotations

@zdtsw can someone from the team enable it?

zdtsw replied:
https://github.com/opendatahub-io/opendatahub-operator/pull/865

bartoszmajsak replied:
LOL I think I was lacking enough caffeine at the time of writing it, I thought it was a repo/action setting somewhere like with the gh tokens.

zdtsw replied:
guess the only way to know if this PR^ works is a rebase :P

bartoszmajsak replied:
> guess the only way to know if this PR^ works is a rebase :P

yeah this is by design, otherwise you could raise malicious PR with elevated privileges and do some weird stuff :)

zdtsw commented on config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:
```diff
@@ -179,6 +179,12 @@ spec:
                       before enable component Does not support enabled ModelMeshServing
                       at the same time
                     properties:
+                      defaultDeploymentMode:
```
you will need to get manifests updates as well.
only change it here will break the build.

VedantMahabaleshwarkar replied:
The original change is written in the Kserve struct. I did not directly edit the CRD. Is there anywhere else I need to define the new field other than the Kserve struct?

zdtsw replied:
no, the code is correct, but to make the build work , we will need catalogsource to update, so the midstream do not break eventually. just run "make" should be enough to update bundles

VedantMahabaleshwarkar replied:
ran make before submitting a new commit. I saw that the `opendatahub-operator/bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml` file is now updated. Hopefully that is what you were talking about?

bartoszmajsak reviewed (changes requested):
I would like to offer some suggestions for further improvement but also have a few questions that can help me understand better its intended use and scope.

I would also very much appreciate implementing some tests, as it would improve maintenance and reliability long term.

Hint: running default `make` target invokes linter so you can see exact errors GH Action complains about right now.

zdtsw commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
 	return nil
 }
 
+func (k *Kserve) setDefaultDeploymentMode(ctx context.Context, cli client.Client, dscispec *dsciv1.DSCInitializationSpec) error {
```
in this case, probably some other methods like configureServerless() can move to a dedicated file as well. 

checkDependentOperators() will not be needed any more, esp. when enabled Raw mode

but all in all, to have this Raw mode on,  all logic related to add/remove serverless  and servicemesh are irrelevant to kserve

VedantMahabaleshwarkar replied:
will try to move these to another file in the next revision

bartoszmajsak commented:
@VedantMahabaleshwarkar if you could rebase your PR (didn't want to do it on your behalf through GH UI) we might see if GH Linter Action will comment on the issues now.

zdtsw commented:
@ykaliuta bought up a good point in this PR: why not we do patch on manifests  instead of changing CM in the cluster.
wont this cause reconcile as manifests drift from CM?

bartoszmajsak commented:
> @ykaliuta bought up a good point in this PR: why not we do patch on manifests instead of changing CM in the cluster.
> wont this cause reconcile as manifests drift from CM?

Very valid point, thanks @ykaliuta (though I do not see your comment here). IIRC, in KfDef version (operator-v1) if the resource already existed it was not reconciled to the "desired" state, it was simply skipped, leaving it customizable to the user. It seems to be different now with server-side apply and forced ownership:

https://github.com/opendatahub-io/opendatahub-operator/blob/45638fdaa235fc836ba99cc9219a14e3c7909c5a/pkg/deploy/deploy.go#L301-L307

This makes all resources part of reconciliation and effectively static. Not sure this is what we want for the configuration resources in particular.

VedantMahabaleshwarkar commented on components/kserve/kserve.go:
```diff
@@ -41,6 +50,9 @@ type Kserve struct {
 	Serving infrav1.ServingSpec `json:"serving,omitempty"`
+	// +kubebuilder:validation:Enum=Serverless;RawDeployment
+	// +kubebuilder:default=Serverless
+	DefaultDeploymentMode string `json:"defaultDeploymentMode,omitempty"`
```
Will add some godoc. 

>Is Serverless used officially to describe the deployment mode in upstream?
Yes, the three options for deployment mode in Kserve are `Serverless`, `RawDeployment`, and `ModelMesh`. However we do not want to present `ModelMesh` as a valid deployment option in our case as we have an entirely separate component for it

ykaliuta replied:
Why it's called default? At the first glance it looks like just mode.

VedantMahabaleshwarkar replied:
It's matching the key used by Kserve in the [configmap ](https://github.com/opendatahub-io/kserve/blob/8f71d3b10b6513deb4795b8739c3ecc9e699e3e8/config/overlays/odh/inferenceservice-config-patch.yaml#L64)for consistency. The mode specified here is the default if nothing is specified in the ISVC. However the ISVC can choose to override the default by using an annotation

VedantMahabaleshwarkar commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
+	if k.DefaultDeploymentMode == RawDeployment {
+		ingressData["ingressClassName"] = RawDeploymentIngressClassName
+	} else {
+		ingressData["ingressClassName"] = IstioIngressClassName
```
The `else` block here is specifically to support going back to `Serverless` from `RawDeployment`. 
Scenario -> 
- Original deployment mode is `Serverless`. Currently Ingress class will be `istio`
- If deployment mode is changed to `RawDeployment`. Now ingress class will be changed to `openshift-default`. 
- If deployment mode is again changed to `Serverless`, we need to change ingress class back to istio

VedantMahabaleshwarkar commented on config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:
```diff
@@ -179,6 +179,12 @@ spec:
+                      defaultDeploymentMode:
+                        default: Serverless
+                        enum:
+                        - Serverless
```
yes the default value is Serverless, which is what we currently have. This configuration will have no change in behavior compared to what exists today. If the defaultDeploymentMode is changed to `RawDeployment` then the configmap is edited accordingly

VedantMahabaleshwarkar commented:
@bartoszmajsak @ykaliuta 

> @ykaliuta bought up a good point in this PR: why not we do patch on manifests instead of changing CM in the cluster.
wont this cause reconcile as manifests drift from CM?
Very valid point, thanks @ykaliuta (though I do not see your comment here). IIRC, in KfDef version (operator-v1) if the resource already existed it was not reconciled to the "desired" state, it was simply skipped, leaving it customizable to the user. It seems to be different now with server-side apply and forced ownership: 

Right now there's only a couple of changes that we are supporting to the `inferenceservice-config` configmap. However if you look at the total number of options presented in the configmap, there's a lot more. IMO relying on manifest patches to generate the correct configmap spec is doable while we are exposing a minimal number of configurable options, but in the future if we decide to expose more options in the DSC it will be difficult to maintain a bunch of manifest patches to support all possible configurations. 
Also, I don't know how we'd be able to support changes made to the DSC after install time using this method. (for e.g: an admin changing the default deployment mode ~2 weeks after install) 

In my testing of this PR, I did not observe repeated reconciliation of the kserve component even though we are diverging from the manifests. I could be wrong here but since we're editing the configmap in the component reconcile loop, once the loop succeeds the operator reports the component as successfully reconciled. The operator would not try to reconcile the configmap back to the original spec according to the manifests here until something else triggers reconciliation again (in which case the configmap will again be set according to the DSC spec)

VedantMahabaleshwarkar commented on components/kserve/kserve.go:
```diff
@@ -166,6 +183,43 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
+	deployDataBytes, _ := json.MarshalIndent(deployData, "", " ")
+	inferenceServiceConfigMap.Data["deploy"] = string(deployDataBytes)
+
+	//if the default deployment mode is RawDeployment, also set the ingressClass to "openshift-default", otherwise it should be "istio"
```
I think the enum you are proposing can be done for the ingress class. For the defaultDeployment enum I was thinking it would be good to match the enum with the enum that Kserve directly expects. This way we can directly pass down the value specified in the DSC, and having the deployment mode enum be consistent might also be less confusing to the user

bartoszmajsak replied:
> I think the enum you are proposing can be done for the ingress class.

Oh yeah, that makes more sense! If you could incorporate it as part of this PR that would be amazing.

> Kserve directly expects

+1000 I missed this aspect :)

VedantMahabaleshwarkar replied:
Based on some ongoing updates in the `kserve` repo. @danielezonca @israel-hdez recommended that instead of setting a value for `IngressClassName` we would rather set `disableIngressCreation` to `true`. Therefore I removed the logic to modify the `IngressClassName`. 
https://github.com/kserve/kserve/pull/3436/files

ykaliuta commented:
Should it include e2e test as well?

bartoszmajsak commented:
> The operator would not try to reconcile the configmap back to the original spec according to the manifests here until something else triggers reconciliation again (in which case the configmap will again be set according to the DSC spec)

This is because the code in question first calls `deploy.DeployManifestsFromPath` and then `setDefaultDeploymentMode` within the same loop. My remark was more as a general principle, as any other config map that does not have such a special treatment will be eventually reverted to the state defined in the provided manifests.

VaishnaviHire commented:
Given this is an API change, tagging @etirelli @danielezonca as well for reviews

israel-hdez reviewed (commented):
This PR seems to be going the right direction, but my thinking tells me the operator can be smarter.

* It looks good to expose the new `defaultDeploymentMode`.
* It is OK to automatically set the ingress class name to `openshift-default`, but I think this field should also be exposed in the DSC. We cannot make the hard assumption that this will always be OK.
* Since the DSCI and the DSC already expose settings for OSSM and Serverless, these new configs can automatically inferred:
  * If in the DSC, the `spec.components.kserve.serving.managementState` field is set to `Removed`, the operator should automatically apply RawDeployment configurations rather than requiring the user to manually specify it (so, we should allow the new field(s) to be empty) -- IMHO, it is redundant to require the user to manually specify the deployment mode, if the operator already has the knowledge that Serverless is going to be absent.
  * Although, if the deployment mode (and, possibly the ingress class) is manually set, then honor whatever the user set.

Additionally, I agree with the comments from others about that patching the ConfigMap like this can lead to a reconciliation loop.

VedantMahabaleshwarkar commented:
In the latest commit, I added a file `kserve_config_handler.go`. I moved the existing functions for dependency setup to this file. I also added logic for a sanitycheck where if `DefaultDeploymentMode = Serverless` but `Serving.ManagementState = Removed`, the operator will throw an error saying that this is incompatible. If `Serving.ManagementState = Removed`, it is assumed that  `DefaultDeploymentMode = RawDeployment`

terrytangyuan commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,148 @@
+	}
+	inferenceServiceConfigMap.Data["deploy"] = string(deployDataBytes)
+
+	// disabling logic to enable or disable ingress creation till https://github.com/kserve/kserve/pull/3436/files is merged
```
Cross reference here: https://github.com/kserve/kserve/pull/3436

bartoszmajsak replied:
Thanks @terrytangyuan. How do we want to coordinate this change? 

Should we wait until your PR is merged and we update KServe manifests version pointing to the new release as part of this PR?

VedantMahabaleshwarkar replied:
@bartoszmajsak I forgot to explicitly mention this change in the latest comment. But the idea here is that after @terrytangyuan's PR is merged, we'd rather set the `disableIngressCreation` field according to the deployment mode rather than edit the `ingressClassName`. The logic will not work till the new field from https://github.com/kserve/kserve/pull/3436 is merged into ODH. We can merge this PR independently and have a follow up PR to uncomment this block later, or we can wait for the other change to make it's way in before merging this PR. I'm okay with both

israel-hdez replied:
@terrytangyuan Do you think KServe would crash if there is a spurious field?
If not, we could commit this code uncommented to avoid the need of a follow-up PR.

terrytangyuan replied:
https://github.com/kserve/kserve/pull/3436 has been merged

israel-hdez replied:
@VedantMahabaleshwarkar would you mind following-up by uncommenting this code?

bartoszmajsak replied:
+1 Otherwise it will leave readers confused seeing this, so that would be great to clean up :)

israel-hdez commented on components/kserve/kserve.go:
```diff
@@ -149,6 +153,11 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
 		}
 	}
+
+	if err := k.setupKserveConfigAndDependencies(ctx, cli, dscispec); err != nil {
```
This moved some logic outside the `if !enabled` block.

This leads to:
* trying to configuring Serverless, and
* trying to apply the patch to configure the default deployment mode

... regardless of KServe management state, which does not looks right.

VedantMahabaleshwarkar replied:
ah, missed this. Fixed by only calling the function if kserve is enabled

VedantMahabaleshwarkar commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,148 @@
+	// since serverless is specifically set to 'Removed'
+	// we do not need to check if dependent operators are installed and configure them
+	// in this case the default deployment mode is forced to be RawDeployment
+	if k.DefaultDeploymentMode == Serverless {
```
If Serverless is set to `Removed` but the default deploymentmode is `Serverless`, I'm not sure if it's better to just throw an error saying that this combination is incompatible or print a warning but go ahead and default to RawDeployment. If we default to RawDeployment in this case, is there a way to set `defaultDeploymentMode` in the DSC to `RawDeployment` ourselves? otherwise internally the deploymentmode will be `RawDeployment` but the DSC will still say `Serverless`, which might be misleading. 
@zdtsw @bartoszmajsak @israel-hdez

israel-hdez replied:
IMHO, if the intent is explicit in the DSC, the right approach is to terminate with error.

VedantMahabaleshwarkar replied:
I would agree with that approach as well. 
@bartoszmajsak @zdtsw if there is consensus I can implement this

zdtsw replied:
does this make it easier to read:
```
switch k.Serving.ManagementState
case operatorv1.Managed, operatorv1.Unmanaged:
	// check on dependent operators if all installed in cluster
	dependOpsErrors := checkDependentOperators(cli).ErrorOrNil()
	if dependOpsErrors != nil {
		return dependOpsErrors
	}

	if err := k.configureServerless(dscispec); err != nil {
		return err
	}
case operatorv1.Removed:
	if k.DefaultDeploymentMode == Serverless {
		return fmt.Errorf("Misconfig: enable Serving in Kserve to use defaultDeployment mode with Serverless")
	} esle {
		fmt.Println("Serverless is specified as 'Removed', Kserve will default to RawDeployment")
	}


if err := k.setDefaultDeploymentMode(ctx, cli, dscispec, k.DefaultDeploymentMode); err != nil {
	return err
}
```

VaishnaviHire replied:
@VedantMahabaleshwarkar @zdtsw Quick question: why are we checking on dependent operators when serving state is `Unmanaged` ?

VedantMahabaleshwarkar replied:
@VaishnaviHire cleaned up logic to only check dependent operators if serving is `Managed`

israel-hdez commented on components/kserve/kserve.go:
```diff
@@ -34,13 +32,28 @@ var (
+	// Configures the default deployment mode for Kserve. This can be set to 'Serverless' or 'RawDeployment'.
+	// The value specified in this field will be used to set the default deployment mode in the 'inferenceservice-config' configmap for Kserve
+	// +kubebuilder:validation:Enum=Serverless;RawDeployment
+	// +kubebuilder:default=Serverless
```
IMO, the default should be empty to:
* be able to have a minimal DSC (i.e. reduce the verbosity);
* and, if empty, this would let the operator choose the best value, given the `managementState` of Serverless (because `managementState` fields don't allow empties).

VedantMahabaleshwarkar replied:
@danielezonca @zdtsw @bartoszmajsak wdyt about this approach?

danielezonca replied:
I agree to avoid "polluting" to much the API so fine for me to have an empty default with the meaning of "Serverless"

VaishnaviHire replied:
So the reason we are not using the pointers in general for fields in DSCI/DSC is there was a previous requirement that if a field is not set, it should still be visible for users with a default value, so that they know readily what configurations are available. [Ref](https://github.com/opendatahub-io/opendatahub-operator/issues/324)

However, this is a field within a component spec, so I think it should be okay to set it as pointer.

VedantMahabaleshwarkar replied:
Removed the `// +kubebuilder:default=Serverless` line to make the default empty and added logic to infer `Serverless` as the default deployment mode if a value is not explicitly provided 
cc @israel-hdez

zdtsw replied:
should we add some description into README for this new feature  (even it wont break nightly)?

israel-hdez replied:
> should we add some description into README for this new feature (even it wont break nightly)?

We, probably, should have a small conversation about where ODH community docs should live. Right now, I'm writing something for KServe authorization and I'm planning to put it on opendatahub-io/kserve repo. I'd cover how to configure the feature (the operator part) and also how to use the feature. Clearly, the how to configure may also belong to this repo. But, then, docs would be scattered.

I'm :+1: on having community docs, but I'm :-1: on not having a single place to host them. At the very least, we should have a single entry point and, then, links (and cross-links) between the relevant repos.

zdtsw commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,148 @@
+
+	// disabling logic to enable or disable ingress creation till https://github.com/kserve/kserve/pull/3436/files is merged
+
+	// var ingressData map[string]interface{}
```
think we can cleanup these uncommented lines

VaishnaviHire commented:
@VedantMahabaleshwarkar This will also need a rebase

VedantMahabaleshwarkar commented:
/retest

israel-hdez commented on components/kserve/kserve.go:
```diff
@@ -144,6 +152,13 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
 	}
+
+	if enabled {
+		if err := k.setupKserveConfigAndDependencies(ctx, cli, dscispec); err != nil {
```
This effectively moves Serverless setup to the end of the KServe reconcile cycle.
I'm not sure that's 100% correct because KServe pods may the have a chance to start before KNative Serving is fully installed, which may lead to KServe to boot-up in non-Serverless mode.

israel-hdez replied:
Additionally, taking into account your changes in opendatahub-io/odh-model-controller#164, I don't think this should be that late in the reconcile cycle.

An upgrade may lead to `odh-model-controller` seeing the wrong values of `defaultDeploymentMode` and lead to reconcile issues. So, right ordering seems to be important:
1. Make sure Serverless is installed, before trying to install KServe.
2. Deploy KServe
3. Make sure the ConfigMap is patched with the right deployment mode before deploying odh-model-controller.
4. Deploy odh-model-controller.

Bullet 4 is not trivial, because ModelMesh may also try to deploy it out-of-order.

israel-hdez replied:
Also, thinking this together with kserve/kserve#3472:

by updating the ConfigMap after KServe is deployed, there is the chance that KServe boots before the ConfigMap is patched. Because of this, as part of bullet 3, I think you will need to restart the KServe pod (if it already exists) to ensure it will read the updated deployment mode.

VedantMahabaleshwarkar replied:
Changed the logic in the latest commit. 
- Moved Serverless setup to earlier in the reconcile loop before Kserve is deployed. 
- Moved Kserve Config setup before odh model controller deployment 
- If the configmap is updated during the reconcile loop, the kserve pod is restarted to reboot with values from the updated configmap
  - If the reconcile loop is triggered but the configmap is not updated, the kserve pod is not restarted

zdtsw replied:
i agree the first 2 points except the last one to have a "kill pod => restart it" inside of this function. 

in this case, the feature is tightly coupled with odh operator.
e.g user can only switch deployment by using this flag via DSC.kserve, (if manually update configmap data, it wont take effects, unless user restart pod later by themselves).
i can accept this as a workaround done by odh operator for now, but in long term, i would think this "kill pod => restart it" logic should be something done by the kserve controller.

israel-hdez replied:
> but in long term, i would think this "kill pod => restart it" logic should be something done by the kserve controller.

I'm not sure if this would be doable, because KServe controller does not have awareness of odh-model-controller. The odh-model-controller is becoming like a glue component for ODH platform.

IMHO, even moving the restart to another controller would be a workaround. What I see more ideal is to re-work the patch as a transformer (which could be a generalized feature in the operator), similar to the Namespace and Labels transformers you already have. Manifests would be modified accordingly and the `manageResource` func would use the final resources, rather than patching after applying. This would fully remove the need of the pod restart.

ykaliuta replied:
> 
>  What I see more ideal is to re-work the patch as a transformer (which could be a generalized feature in the operator), similar to the Namespace and Labels transformers you already have. Manifests would be modified accordingly and the `manageResource` func would use the final resources, rather than patching after applying. This would fully remove the need of the pod restart... but, well, I think this would be a enhancement for another PR - we could coordinate to make this near term.

And that was originally proposed :)

israel-hdez replied:
@ykaliuta Ah, interesting! I guess it is related to [this comment](https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1943582932), right?

I didn't find your original comment, so I didn't have the full context and never followed up. Well... you can blame me about that :sweat_smile:

ykaliuta replied:
> @ykaliuta Ah, interesting! I guess it is related to [this comment](https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1943582932), right?
> 
> I didn't find your original comment, so I didn't have the full context and never followed up. Well... you can blame me about that 😅

Right, but one comment above, this https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1943582932 and originally it was a conversation off-thread. No blaming at all. The conclusion came there.  So there are pros and cons https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1944124092

VedantMahabaleshwarkar commented:
/retest

israel-hdez commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,163 @@
+		case operatorv1.Unmanaged, operatorv1.Removed:
+			return fmt.Errorf("ServiceMesh is need to set to 'Managed' in DSCI CR, it is required by KServe serving field")
+		}
+
```
A call to `checkDependentOperators` is missing here. Looks like a rebase mistake.

VedantMahabaleshwarkar replied:
should be fixed now

israel-hdez commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,163 @@
+		}
+
+	case operatorv1.Unmanaged:
+		fmt.Println("Serverless is Unmanaged, Kserve will default to RawDeployment")
```
AFAIK, `Unmanaged` would mean that the users are bringing their own Serverless setup. Right, @VaishnaviHire @zdtsw ?

If so, both `Managed` and `Unmanaged` should operate identically wrt the logic around `defaultDeploymentMode`.

israel-hdez commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,163 @@
+
+	switch k.Serving.ManagementState {
+	case operatorv1.Managed:
+		// check on dependent operators if all installed in cluster
+		dependOpsErrors := checkDependentOperators(cli).ErrorOrNil()
+		if dependOpsErrors != nil {
+			return dependOpsErrors
+		}
```
This looks also related to my other comment about the rebase mistake.

It may have been the case that the call to `checkDependentOperators` was moved around in a previously merged PR, and may no longer relevant here (because `configureServerless` would already do the call when needed).

VedantMahabaleshwarkar replied:
Fixed

israel-hdez commented on components/kserve/kserve_config_handler.go:
```diff
@@ -0,0 +1,163 @@
+			return dependOpsErrors
+		}
+
+		if err := k.configureServerless(dscispec); err != nil {
+			return err
+		}
```
I don't think it is right to have the call to `configureServerless` inside the `switch` block, given the original code does not have any conditions around `Serving.ManagementState` for invoking this func.

The unwanted side-effect I can see is that flipping `Serving.ManagementState` from `Managed` to `Removed` won't lead to the removal of Knative.

VedantMahabaleshwarkar replied:
moved `configureServerless` out of this function and directly called it in the Kserve reconcile loop before Kserve is deployed.

VedantMahabaleshwarkar commented on components/kserve/kserve.go:
```diff
@@ -34,13 +32,27 @@ var (
+	// Configures the default deployment mode for Kserve. This can be set to 'Serverless' or 'RawDeployment'.
+	// The value specified in this field will be used to set the default deployment mode in the 'inferenceservice-config' configmap for Kserve
+	// +kubebuilder:validation:Enum=Serverless;RawDeployment
+	DefaultDeploymentMode DefaultDeploymentMode `json:"defaultDeploymentMode,omitempty"`
```
I tested the PR and did not observe any failures due to the empty string 
![image](https://github.com/opendatahub-io/opendatahub-operator/assets/20891020/04a060f9-22c3-4ae1-85dd-50e931143c9b)

zdtsw replied:
since we are introducing "" now , should we add some into description like "if not set defaultDeploymentMode, it falls back to Serverless mode"?

VedantMahabaleshwarkar replied:
added description

israel-hdez commented:
This
/lgtm

I'll let the operator team to do their review and give the approval to merge.

VedantMahabaleshwarkar commented:
@zdtsw @VaishnaviHire can we merge this? We have an lgtm from the serving team

zdtsw reviewed (approved):
/lgtm

zdtsw commented:
/lgtm

danielezonca commented:
@VedantMahabaleshwarkar 
Sorry for the last minute request but I would like to expose in the DSC the option `disableIngressCreation` (next to the `DefaultDeploymentMode` option) that has been added by @terrytangyuan ( https://github.com/kserve/kserve/pull/3436/files ).
The reason is that Ingress creation is "broken" (it requires a lot of conf that we don't support yet) and it is tied to RawDeployment but it cannot be "inferred" based on `defaultDeploymentMode` because user can specify the deploymentMode even directly in the `InferenceService`.

Long story short I would like to have this option available in DSC for configuration but with empty default value so it doesn't pollute the example
